File: problem3.java
//time complexity-O(mxn)
//space complexity-O(1)
//passed all test cases on leetcode
//the approach here is to traverse the matrix using the 4 pointers left,top,right,bottom.
class Solution {
    public List<Integer> spiralOrder(int[][] matrix) {
        int m = matrix.length;
        int n = matrix[0].length;
        List<Integer> al = new ArrayList<>();
        int left = 0, top = 0, right = n - 1, bottom = m - 1;
        while (left <= right && top <= bottom) {
            for (int i = left; i <= right; i++) {
                al.add(matrix[top][i]);
            }
            top++;
            for (int i = top; i <= bottom; i++) {
                al.add(matrix[i][right]);
            }
            right--;
            if (top <= bottom) {
                for (int i = right; i >= left; i--) {
                    al.add(matrix[bottom][i]);
                }
                bottom--;
            }
            if (left <= right) {
                for (int i = bottom; i >= top; i--) {
                    al.add(matrix[i][left]);
                }
                left++;
            }
        }
        return al;
    }
}

File: Problem3.py
#T.C. O(m*n) where m = no of rows, n = no of columns
#S.C. O(n) due to result array

class Solution:
    def spiralOrder(self, matrix: List[List[int]]) -> List[int]:
        row = len(matrix)
        col = len(matrix[0])
        top = 0
        bottom = row - 1
        left = 0
        right = col - 1
        result = []
        while left <= right and top <= bottom:
            #top
            for j in range(left, right+1):
                result.append(matrix[top][j])

            top += 1
             #right
            for i in range(top, bottom + 1):
                result.append(matrix[i][right])
            right -= 1
            
            if not (left <= right and top <= bottom):
                break

            #bottom
            for j in range(right, left - 1, -1):
                result.append(matrix[bottom][j])
            bottom -= 1

            #left
            for i in range(bottom, top - 1, -1):
                result.append(matrix[i][left])
            left += 1

        return result
            
